src/class/bth/bth_device.c

```diff
@@ -0,0 +1,238 @@
+
+    TU_ASSERT(TUSB_DESC_ENDPOINT == desc_ep->bDescriptorType && TUSB_XFER_INTERRUPT == desc_ep->bmAttributes.xfer);
+    TU_ASSERT(dcd_edpt_open(rhport, desc_ep));
+    _btd_itf.ep_ev = desc_ep->bEndpointAddress;
+
+    // Open endpoint pair
+    TU_ASSERT(usbd_open_edpt_pair(rhport, tu_desc_next(desc_ep), 2, TUSB_XFER_BULK, &_btd_itf.ep_acl_out,
+                                  &_btd_itf.ep_acl_in));
+
+    // Prepare for incoming data from host
+    TU_ASSERT(usbd_edpt_xfer(rhport, _btd_itf.ep_acl_out, _btd_itf.epout_buf, CFG_TUD_BTH_DATA_EPSIZE), false);
+
+    (*p_length) += sizeof(tusb_desc_interface_t) + 3 * sizeof(tusb_desc_endpoint_t);
+  }
+  else if (itf_desc->bNumEndpoints == 2)
+  {
+    uint8_t dir;
+
+    desc_ep = (tusb_desc_endpoint_t const *) tu_desc_next(itf_desc);
+    TU_ASSERT(desc_ep->bDescriptorType == TUSB_DESC_ENDPOINT);
+    dir = tu_edpt_dir(desc_ep->bEndpointAddress);
+    _btd_itf.ep_voice[dir] = desc_ep->bEndpointAddress;
+    // Store endpoint size for alternative 0
+    _btd_itf.ep_voice_size[dir][0] = (uint8_t) desc_ep->wMaxPacketSize.size;
+
+    desc_ep = (tusb_desc_endpoint_t const *) tu_desc_next(desc_ep);
+    TU_ASSERT(desc_ep->bDescriptorType == TUSB_DESC_ENDPOINT);
+    dir = tu_edpt_dir(desc_ep->bEndpointAddress);
+    _btd_itf.ep_voice[dir] = desc_ep->bEndpointAddress;
+    // Store endpoint size for alternative 0
+    _btd_itf.ep_voice_size[dir][0] = (uint8_t) desc_ep->wMaxPacketSize.size;
+
+    for (int i = 1; i < CFG_TUD_BTH_ISO_ALT_COUNT; ++i) {
+      // Make sure rest of alternatives matches
+      itf_desc = (tusb_desc_interface_t const *)tu_desc_next(desc_ep);
+      TU_ASSERT(itf_desc->bDescriptorType == TUSB_DESC_INTERFACE);
+      TU_ASSERT(TUSB_CLASS_WIRELESS_CONTROLLER == itf_desc->bInterfaceClass &&
+                TUD_BT_APP_SUBCLASS == itf_desc->bInterfaceSubClass &&
+                TUD_BT_PROTOCOL_PRIMARY_CONTROLLER == itf_desc->bInterfaceProtocol);
+
+      desc_ep = (tusb_desc_endpoint_t const *)tu_desc_next(itf_desc);
+      dir = tu_edpt_dir(desc_ep->bEndpointAddress);
+      // Verify that alternative endpoint are same as first ones
+      TU_ASSERT(desc_ep->bDescriptorType == TUSB_DESC_ENDPOINT &&
+                _btd_itf.ep_voice[dir] == desc_ep->bEndpointAddress);
+      _btd_itf.ep_voice_size[dir][i] = (uint8_t) desc_ep->wMaxPacketSize.size;
+
+      desc_ep = (tusb_desc_endpoint_t const *)tu_desc_next(desc_ep);
+      dir = tu_edpt_dir(desc_ep->bEndpointAddress);
+      // Verify that alternative endpoint are same as first ones
+      TU_ASSERT(desc_ep->bDescriptorType == TUSB_DESC_ENDPOINT &&
+                _btd_itf.ep_voice[dir] == desc_ep->bEndpointAddress);
+      _btd_itf.ep_voice_size[dir][i] = (uint8_t) desc_ep->wMaxPacketSize.size;
+    }
+    // Consume all alternatives descriptors
+    (*p_length) += CFG_TUD_BTH_ISO_ALT_COUNT * (sizeof(tusb_desc_interface_t) + 2 * sizeof(tusb_desc_endpoint_t));
```

Comment:
although most users will use descriptor template, but it is totally up to application to define their own configuration descriptors, the `CFG_TUD_BTH_ISO_ALT_COUNT` should be treated as **maximum** number of ALT instead of **exact** number, primarily for storing alternate EPSize. We should still count the actual dynamic alternate interface. **It is probably better to increase `p_length` within the above loop()**

Reply:
I understand your concern.
There is one problem though.
If ```CFG_TUD_BTH_ISO_ALT_COUNT``` was a maximum ```btd_open``` would need to check during runtime if descriptor is valid and should be used to increase ```p_length```.
In this case however function does not know how much memory to check since there is not upper limit provided. So it may end up checking memory that does not contain configuration descriptor.
If open function had limit it would be enough.
I guess that other interfaces do not have this dynamic size nature.

On the other hand if alternate interfaces were handled in function that calls open it could be also covered.

Reply:
@kasjer ah, I see, the driver open() should pass the remaining length left from config descriptor to the open(). Don't worry about that for now, I will create an separated issue for it.

Reply:
Done, waits to fix for out of bounds access

Reply:
implemented by pr #418 and merge, please pull from master to update

Reply:
Update to reflect changes in #418